design/one-pager-generic-usage-type.md

```diff
@@ -0,0 +1,397 @@
+				},
+			},
+		}
+	}
+	return admission.Allowed("")
+}
+```
+
+### User Experience
+
+**Defining a Usage Resource as part of a Composition:**
+
+We expect a typical usage relationship to be defined as part of the same
+composition as the using and used resources. In this case, the user will
+define the `Usage` resource as part of the composition template.
+
+Let's take [upbound/platform-ref-aws] as an example. In this composition, we
+have a [composition for `XCluster`] which composes `XEKS` composing an EKS cluster,
+and `XServices` composing a set of helm releases running on the cluster. In this
+case, we will add the following resource to the same composition to define a
+usage relationship between the `XServices` and the `XEKS` Composites:
+
+```yaml
+    - name: xservices-uses-xeks
+      base:
+        apiVersion: apiextensions.crossplane.io/v1alpha1
+        kind: Usage
+        spec:
+          of:
+            - apiVersion: aws.platformref.upbound.io/v1alpha1
+              kind: XEKS
+              selector:
+                matchControllerRef: true
+          by:
+            apiVersion: aws.platformref.upbound.io/v1alpha1
+            kind: XServices
+            selector:
+              matchControllerRef: true
+```
+
+**Deleting a Claim or Composite which contains a Usage Relationship:**
+
+**With the (default) background propagation policy**, we expect no different in user
+experience. Users will be able to delete the Claim or Composite without any error. 
+The Claim or Composite will be deleted immediately, and the Kubernetes garbage
+collector will delete the Composed Resources together with the `Usage` resources
+in the background. It will take longer until all Composed Resources are garbage
+collected, however this will not be visible to the user.
+
+**With the foreground propagation policy**, the only difference we expect is
+that the deletion is taking longer than before. The user will be able to delete
+the Claim or Composite without any error. The deletion of the Claim or Composite
+will be blocked until all Composed Resources are deleted as before. It will take
+longer because deletion of the used resource will be blocked until the used
+resource hence `Usage` is deleted.
+
+> One caveat with the garbage collector is the substantial delay it can cause
+during the deletion of composites. This delay caused by the garbage collector's
+use of an exponential backoff when the DELETE API call is rejected. The maximum
+backoff time is 1000 seconds, which means that after no usage of an object is
+left, it could take up to 1000 seconds before the garbage collector attempts to
+delete the resource that was previously blocked. To mitigate this issue, we
+could introduce a mechanism in the controller to re-trigger the deletion of the
+used resource as soon as the Usage resource is deleted.
```

Comment:
TBH, I was under the impression that we could just reset the backoff interval of the GC by making some modifications on the resource, e.g. adding an annotation. This is how one could retrigger a typical k8s controller watching a resource bypassing the backoff time so that the reconciler logic could rerun. I thought this would work the same with the Garbage Collector, I need to check that 🤞 or need to see if there are other ways to do so.
If not, I would be hesitant to store and trigger a deletion when the usage disappears, which could cause headaches due to different edge cases and result in a not-so-pleasant user experience for other clients. For example, a user gets rejected by trying to delete a used resource and thinks deletion **didn't happen**, but we do that behind the scenes.

Reply:
Thank you @sttts, that trick was something I also found accidentally :) Obviously, adding an invalid owner ref is hacky but that sparked another idea.

What we want to achieve here is ideally the Garbage Collector not to attempt deleting a resource that has at least one Usage and only to do that when there is no Usage. One obvious way to achieve this is adding an owner reference from `Usage` to the used resource **as long as the Usage exists** (This owner reference will be cleaned up during the finalization of the `Usage` resource). In the case of composed resources, this _additional_ owner reference will prevent the resource to be Garbage collected, and once the Usage is gone, the resource will become GCable. This would mean no backoff during Garbage Collection.

We are already planning to add owner references from Using resource to the `Usage`, and with the above, it would look like:

![image](https://github.com/crossplane/crossplane/assets/9900707/2de1a4c9-7784-40c2-967d-3987de64c8ad)

Reply:
`blockOwnerDeletion` will be `true` on the ownerReference from the `Usage` to the `Cluster`?

Does this also block the deletion of the `Cluster`'s other child objects?

Reply:
> `blockOwnerDeletion` will be `true` on the ownerReference from the `Usage` to the `Cluster`?

I don't see an immediate reason to set `blockOwnerDeletion` to `true` there.

> Does this also block the deletion of the Cluster's other child objects when using foreground cascading deletion?

I think so. Whether it is background or foreground, I would expect GC process not to start as long as there is at least one existing owner. But I will keep this in mind during implementation and make sure that it works as expected.

Reply:
> One obvious way to achieve this is adding an owner reference from Usage to the used resource as long as the Usage exists (This owner reference will be cleaned up during the finalization of the Usage resource)

One caveat I noticed during implementation is that it could cause unexpected deletion of `used` resource if someone _manually_ removes the finalizer on the `Usage` resource since owner ref from `Usage` to `used` resource is not cleaned up. I would be ok to this by saying "one should never manually remove a finalizer; otherwise unexpected things could happen", however, there is at least one use case , which I believe, we would need manual removal of finalizer on the `Usage`, which is deleting the `Usage` while `using` resource is still there.

Imagine I created a `Release`, `Cluster`, and a `Usage` as `release-using-cluster` as in the above figure. Since both the `Release` and `Usage` could be part of the same composition, we propose using finalizers to block the deletion of `Usage` before `Release` disappears. So, a `Usage` would never be deleted unless `Release` is gone. However, there could be some cases where someone changed their mind and want to just clean up the `Usage` just to break the dependency. Removing the finalizer would be a straightforward solution in this case unless we implement additional knobs like an annotation to force delete etc.

So, I am on the fence regarding implementing this workaround to speed up garbage collector as it could be fatal (used resource could be deleted) if someone messes up with finalizers.

Reply:
> (used resource could be deleted) if someone messes up with finalizers.

I think (or I hope) most people understand that removing finalizers is at-your-own-risk.  I like using finalizers better than re-triggering the delete call, and the controller can maintain the finalizer until the `Release` (in this case) has been deleted.  I think it's worth a try if the only alternative is waiting for GC to come around again in 17m.

Reply:
Testing this with Compositions and unfortunately observed a behaviour that invalidates it as a solution :disappointed:

**TL;DR:** If the Garbage collector finds an owner to a nonexisting resource, it removes that owner from the owner list if it is not the only owner (which would otherwise delete the resource). Due to this, we are losing owner ref to composite before we clean up the one for Usage, hence we end up with a non GC-ed resource with no owner ref.

Here is what happened:

1. Created a composite composing `used` and `using` resources together with the `Usage` itself.
2. Composite controller created `used` resource with owner reference to itself.
3. The `Usage` controller added owner reference to the `used` resource for itself, as proposed above.

```yaml
metadata:
  labels:
    usage: used
  name: test-hwgrp-pj72k
  ownerReferences:
  - apiVersion: nop.example.org/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: XNopResource
    name: test-hwgrp
    uid: 697b0b9e-0a41-48ea-b624-007bbf444175
  - apiVersion: apiextensions.crossplane.io/v1alpha1
    kind: Usage
    name: test-hwgrp-g55x8
    uid: 35975f94-d3dd-444e-8095-c4894f81a089
```

So far, so good, as planned.

4. Deleted the composite, and I ended up with the `used` resource not deleted without any ownerReferences. After some thinking, I think this is expected, and here is what happens:
    1. The composite is deleted.
    2. On the used resource, the garbage collector finds an owner ref to a non-existing resource, but it is **not** the only owner ref. So, removes the owner ref to the composite from the owner ref list but do not delete the resource since there is another owner, i.e. the Usage resource.
    3. The Usage controller acts after GC and removes its owner ref.
    4. We end up an orphaned composed resource without any owner ref.

Reply:
My plan for now:

I'll focus on implementing the first version of my PR without this and put more thought and look for alternative solutions (like [tricks to re-queue for GC](https://github.com/crossplane/crossplane/pull/4215#discussion_r1238738203)) later.